dadi/lib/controller/index.js

```diff
@@ -343,6 +348,48 @@ Controller.prototype.stats = function (req, res, next) {
   })
 }
 
+/**
+ * Handle collection search endpoints
+ * Example: /1.0/library/books/search?q=title
+ */
+Controller.prototype.search = function (req, res, next) {
+  let path = url.parse(req.url, true)
+  let options = path.query
+
+  let queryOptions = this._prepareQueryOptions(options)
+
+  if (queryOptions.errors.length !== 0) {
+    sendBackJSON(400, res, next)(null, queryOptions)
+  } else {
+    queryOptions = queryOptions.queryOptions
+  }
+
+  return this.model.search(queryOptions).then(query => {
+    let ids = query._id['$containsAny'].map(id => id.toString())
+
+    return this.model.get({
+      query,
+      options: queryOptions,
+      req
+    }).then(results => {
+      results.results = results.results.sort((a, b) => {
+        let aIndex = ids.indexOf(a._id.toString())
+        let bIndex = ids.indexOf(b._id.toString())
+
+        if (aIndex === bIndex) return 0
+
+        return aIndex > bIndex ? 1 : -1
+      })
+
+      return help.sendBackJSON(200, res, next)(null, results)
+    }).catch(error => {
```

Comment:
I believe you don't need this `catch`, since you have one on the parent Promise.